File: web/src/types/wallet.ts
// 钱包类型
export type WalletType = 'profit' | 'service' | 'reward' | 'recharge' | 'deposit'

// 钱包信息
export interface Wallet {
  id: number
  agent_id: number
  agent_name: string
  channel_id: number
  channel_name: string
  wallet_type: WalletType
  balance: number
  frozen: number
  available: number
  total_income: number
  total_withdraw: number
  updated_at: string
}

// 钱包流水
export interface WalletLog {
  id: number
  wallet_id: number
  log_no: string
  type: 'income' | 'expense' | 'freeze' | 'unfreeze' | 'withdraw'
  amount: number
  balance_before: number
  balance_after: number
  related_no: string
  remark: string
  created_at: string
}

// 钱包汇总
export interface WalletSummary {
  profit_balance: number
  service_balance: number
  reward_balance: number
  recharge_balance: number
  deposit_balance: number
  total_balance: number
  total_available: number
  total_frozen: number
}

// 钱包类型配置
export const WALLET_TYPE_CONFIG: Record<WalletType, { label: string; color: string; description: string }> = {
  profit: { label: '分润钱包', color: '#409eff', description: '交易分润收入' },
  service: { label: '服务费钱包', color: '#67c23a', description: '流量费+押金返现' },
  reward: { label: '奖励钱包', color: '#e6a23c', description: '激活奖励收入' },
  recharge: { label: '充值钱包', color: '#f56c6c', description: '上级额外奖励' },
  deposit: { label: '沉淀钱包', color: '#909399', description: '下级未提现余额' },
}

// 提现申请
export interface WithdrawRequest {
  wallet_id: number
  amount: number
}

// 提现记录
export interface Withdrawal {
  id: number
  withdrawal_no: string
  agent_id: number
  agent_name: string
  channel_id: number
  channel_name: string
  wallet_type: WalletType
  amount: number
  fee: number
  actual_amount: number
  bank_name: string
  bank_card_no: string
  status: 'pending' | 'approved' | 'rejected' | 'paid' | 'failed'
  remark: string
  created_at: string
  audited_at: string
  paid_at: string
}

// 钱包流水查询参数
export interface WalletLogQueryParams {
  wallet_id?: number
  type?: string
  start_date?: string
  end_date?: string
  page?: number
  page_size?: number
}

// ========== 充值钱包相关类型 ==========

// 代理商钱包配置
export interface AgentWalletConfig {
  agent_id: number
  charging_wallet_enabled: boolean
  charging_wallet_limit: number
  charging_wallet_limit_yuan: number
  settlement_wallet_enabled: boolean
  settlement_ratio: number
  enabled_at?: string
}

// 开通充值钱包请求
export interface EnableChargingWalletParams {
  agent_id: number
  limit: number // 限额(分)
}

// 充值钱包汇总
export interface ChargingWalletSummary {
  balance: number
  balance_yuan: number
  total_issued: number
  total_issued_yuan: number
}

// 充值记录状态
export type ChargingDepositStatus = 0 | 1 | 2 // 0=待确认 1=已确认 2=已拒绝

// 充值记录
export interface ChargingWalletDeposit {
  id: number
  deposit_no: string
  agent_id: number
  amount: number
  amount_yuan: number
  payment_method: number
  payment_method_name: string
  payment_ref: string
  status: ChargingDepositStatus
  status_name: string
  reject_reason?: string
  remark?: string
  created_at: string
  confirmed_at?: string
}

// 申请充值请求
export interface CreateChargingDepositParams {
  amount: number // 分
  payment_method: number // 1=银行转账 2=微信 3=支付宝
  payment_ref?: string
  remark?: string
}

// 奖励发放记录
export interface ChargingWalletReward {
  id: number
  reward_no: string
  from_agent_id: number
  from_agent_name: string
  to_agent_id: number
  to_agent_name: string
  amount: number
  amount_yuan: number
  reward_type: number
  reward_type_name: string
  status: number
  status_name: string
  remark?: string
  created_at: string
  revoked_at?: string
}

// 发放奖励请求
export interface IssueChargingRewardParams {
  to_agent_id: number
  amount: number // 分
  remark?: string
}

// 充值记录查询参数
export interface ChargingDepositQueryParams {
  status?: ChargingDepositStatus
  page?: number
  page_size?: number
}

// 奖励记录查询参数
export interface ChargingRewardQueryParams {
  direction: 'from' | 'to' // from=我发放的, to=我收到的
  page?: number
  page_size?: number
}

// ========== 沉淀钱包相关类型 ==========

// 开通沉淀钱包请求
export interface EnableSettlementWalletParams {
  agent_id: number
  ratio: number // 沉淀比例(1-100)
}

// 沉淀钱包汇总
export interface SettlementWalletSummary {
  subordinate_total_balance: number
  subordinate_total_balance_yuan: number
  settlement_ratio: number
  available_amount: number
  available_amount_yuan: number
  used_amount: number
  used_amount_yuan: number
  pending_return_amount: number
  pending_return_amount_yuan: number
}

// 下级余额信息
export interface SubordinateBalance {
  agent_id: number
  agent_name: string
  available_balance: number
  available_balance_yuan: number
}

// 沉淀使用类型
export type SettlementUsageType = 1 | 2 // 1=使用 2=归还

// 沉淀使用状态
export type SettlementUsageStatus = 1 | 2 // 1=正常 2=待归还

// 沉淀使用记录
export interface SettlementWalletUsage {
  id: number
  usage_no: string
  agent_id: number
  amount: number
  amount_yuan: number
  usage_type: SettlementUsageType
  usage_type_name: string
  status: SettlementUsageStatus
  status_name: string
  return_deadline?: string
  returned_at?: string
  remark?: string
  created_at: string
}

// 使用沉淀款请求
export interface UseSettlementParams {
  amount: number // 分
  remark?: string
}

// 归还沉淀款请求
export interface ReturnSettlementParams {
  amount: number // 分
  remark?: string
}

// 沉淀使用记录查询参数
export interface SettlementUsageQueryParams {
  usage_type?: SettlementUsageType
  page?: number
  page_size?: number
}

// ========== 钱包拆分相关类型 ==========

// 代理商钱包拆分配置
export interface AgentWalletSplitConfig {
  id?: number
  agent_id: number
  split_by_channel: boolean
  configured_by?: number
  configured_at?: string
  created_at?: string
  updated_at?: string
}

// 政策模版提现门槛配置
export interface PolicyWithdrawThreshold {
  id?: number
  template_id: number
  wallet_type: number // 1=分润 2=服务费 3=奖励
  channel_id: number // 0=通用
  threshold_amount: number // 分
  created_at?: string
  updated_at?: string
}

// 子钱包（按通道拆分时的二级钱包）
export interface SubWallet {
  channel_id: number
  channel_name: string
  balance: number
  frozen_amount: number
  withdraw_threshold: number
  can_withdraw: boolean
}

// 钱包展示（支持拆分模式）
export interface WalletDisplay {
  wallet_type: number
  wallet_type_name: string
  balance: number
  frozen_amount: number
  total_income: number
  total_withdraw: number
  withdraw_threshold: number
  can_withdraw: boolean
  sub_wallets?: SubWallet[]
}

// 钱包列表响应（支持拆分模式）
export interface WalletListWithSplitResponse {
  split_by_channel: boolean
  wallets: WalletDisplay[]
}

// 设置拆分配置请求
export interface SetSplitConfigRequest {
  split_by_channel: boolean
}

// 设置提现门槛请求
export interface SetWithdrawThresholdRequest {
  wallet_type: number
  channel_id: number
  threshold_amount: number
}

// 带通道的提现请求
export interface WithdrawWithChannelRequest {
  wallet_id: number
  channel_id?: number // 拆分模式下必填
  amount: number
}

File: server/internal/models/wallet.go
package models

import (
	"time"
)

// 注意: 钱包类型常量定义在 policy.go 中
// WalletTypeProfit     int16 = 1 // 分润钱包
// WalletTypeService    int16 = 2 // 服务费钱包
// WalletTypeServiceFee int16 = 2 // 服务费钱包（别名）
// WalletTypeReward     int16 = 3 // 奖励钱包
// WalletTypeCharging   int16 = 4 // 充值钱包
// WalletTypeSettlement int16 = 5 // 沉淀钱包

// GetWalletTypeName 获取钱包类型名称（使用 WalletTypeName）
func GetWalletTypeName(walletType int16) string {
	return WalletTypeName(walletType)
}

// AgentWalletConfig 代理商特殊钱包配置
type AgentWalletConfig struct {
	ID int64 `json:"id" gorm:"primaryKey"`

	AgentID int64 `json:"agent_id" gorm:"uniqueIndex"`

	// 充值钱包配置
	ChargingWalletEnabled bool  `json:"charging_wallet_enabled" gorm:"default:false"`
	ChargingWalletLimit   int64 `json:"charging_wallet_limit" gorm:"default:0"` // 充值钱包限额(分)

	// 沉淀钱包配置
	SettlementWalletEnabled bool `json:"settlement_wallet_enabled" gorm:"default:false"`
	SettlementRatio         int  `json:"settlement_ratio" gorm:"default:30"` // 沉淀比例(百分比)

	// 审计字段
	EnabledBy *int64     `json:"enabled_by"`
	EnabledAt *time.Time `json:"enabled_at"`
	CreatedAt time.Time  `json:"created_at" gorm:"default:now()"`
	UpdatedAt time.Time  `json:"updated_at" gorm:"default:now()"`
}

// TableName 表名
func (AgentWalletConfig) TableName() string {
	return "agent_wallet_configs"
}

// 充值记录状态
const (
	ChargingDepositStatusPending   int16 = 0 // 待确认
	ChargingDepositStatusConfirmed int16 = 1 // 已确认
	ChargingDepositStatusRejected  int16 = 2 // 已拒绝
)

// GetChargingDepositStatusName 获取充值状态名称
func GetChargingDepositStatusName(status int16) string {
	switch status {
	case ChargingDepositStatusPending:
		return "待确认"
	case ChargingDepositStatusConfirmed:
		return "已确认"
	case ChargingDepositStatusRejected:
		return "已拒绝"
	default:
		return "未知"
	}
}

// ChargingWalletDeposit 充值钱包充值记录
type ChargingWalletDeposit struct {
	ID            int64  `json:"id" gorm:"primaryKey"`
	DepositNo     string `json:"deposit_no" gorm:"size:50;uniqueIndex"`
	AgentID       int64  `json:"agent_id" gorm:"index"`
	Amount        int64  `json:"amount"`                        // 充值金额(分)
	PaymentMethod int16  `json:"payment_method" gorm:"default:1"` // 1=银行转账 2=微信 3=支付宝
	PaymentRef    string `json:"payment_ref" gorm:"size:100"`     // 支付流水号
	Status        int16  `json:"status" gorm:"default:0"`

	// 审核信息
	ConfirmedBy  *int64     `json:"confirmed_by"`
	ConfirmedAt  *time.Time `json:"confirmed_at"`
	RejectReason string     `json:"reject_reason" gorm:"size:500"`

	Remark    string    `json:"remark" gorm:"size:500"`
	CreatedBy int64     `json:"created_by"`
	CreatedAt time.Time `json:"created_at" gorm:"default:now()"`
	UpdatedAt time.Time `json:"updated_at" gorm:"default:now()"`
}

// TableName 表名
func (ChargingWalletDeposit) TableName() string {
	return "charging_wallet_deposits"
}

// 奖励类型
const (
	ChargingRewardTypeManual int16 = 1 // 手动发放
	ChargingRewardTypeAuto   int16 = 2 // 自动发放(政策触发)
)

// 奖励状态
const (
	ChargingRewardStatusIssued  int16 = 1 // 已发放
	ChargingRewardStatusRevoked int16 = 2 // 已撤销
)

// ChargingWalletReward 充值钱包奖励发放记录
type ChargingWalletReward struct {
	ID          int64 `json:"id" gorm:"primaryKey"`
	RewardNo    string `json:"reward_no" gorm:"size:50;uniqueIndex"`
	FromAgentID int64 `json:"from_agent_id" gorm:"index"`
	ToAgentID   int64 `json:"to_agent_id" gorm:"index"`
	Amount      int64 `json:"amount"`                          // 奖励金额(分)
	RewardType  int16 `json:"reward_type" gorm:"default:1"`    // 1=手动发放 2=自动发放
	PolicyID    *int64 `json:"policy_id"`                       // 关联的奖励政策ID

	// 状态
	Status       int16      `json:"status" gorm:"default:1"`
	RevokedAt    *time.Time `json:"revoked_at"`
	RevokeReason string     `json:"revoke_reason" gorm:"size:500"`

	Remark    string    `json:"remark" gorm:"size:500"`
	CreatedBy int64     `json:"created_by"`
	CreatedAt time.Time `json:"created_at" gorm:"default:now()"`
}

// TableName 表名
func (ChargingWalletReward) TableName() string {
	return "charging_wallet_rewards"
}

// GetRewardTypeName 获取奖励类型名称
func GetRewardTypeName(rewardType int16) string {
	switch rewardType {
	case ChargingRewardTypeManual:
		return "手动发放"
	case ChargingRewardTypeAuto:
		return "自动发放"
	default:
		return "未知"
	}
}

// GetRewardStatusName 获取奖励状态名称
func GetRewardStatusName(status int16) string {
	switch status {
	case ChargingRewardStatusIssued:
		return "已发放"
	case ChargingRewardStatusRevoked:
		return "已撤销"
	default:
		return "未知"
	}
}

// 沉淀钱包使用类型
const (
	SettlementUsageTypeUse    int16 = 1 // 使用
	SettlementUsageTypeReturn int16 = 2 // 归还
)

// 沉淀钱包使用状态
const (
	SettlementUsageStatusNormal    int16 = 1 // 正常
	SettlementUsageStatusToReturn  int16 = 2 // 待归还
)

// SettlementWalletUsage 沉淀钱包使用记录
type SettlementWalletUsage struct {
	ID        int64  `json:"id" gorm:"primaryKey"`
	UsageNo   string `json:"usage_no" gorm:"size:50;uniqueIndex"`
	AgentID   int64  `json:"agent_id" gorm:"index"`
	Amount    int64  `json:"amount"`                       // 使用金额(分)
	UsageType int16  `json:"usage_type" gorm:"default:1"`  // 1=使用 2=归还

	// 来源明细(哪些下级的钱)
	SourceDetails string `json:"source_details" gorm:"type:jsonb"`

	// 状态
	Status         int16      `json:"status" gorm:"default:1"`
	ReturnDeadline *time.Time `json:"return_deadline"`
	ReturnedAt     *time.Time `json:"returned_at"`

	Remark    string    `json:"remark" gorm:"size:500"`
	CreatedBy int64     `json:"created_by"`
	CreatedAt time.Time `json:"created_at" gorm:"default:now()"`
}

// TableName 表名
func (SettlementWalletUsage) TableName() string {
	return "settlement_wallet_usages"
}
